### PR TITLE
feat: add infinite scrolling to ADataTable

### PR DESCRIPTION
**Pull request by brennarvo**
**Please check if the PR fulfills these requirements**
- [x] The commit message follows semantic commit message guidelines
- [x] The changes are documented in component docs and changelog
- [x] The ESLint plugin has been updated if a new component is added
- [x] Test have been added or modified, if appropriate
- [x] Has been verified locally

**What kind of change does this PR introduce?** 

* Introduces a new `AInView` component that determines if its child component is currently visible within a specified scrollable container (such as the browser viewport itself, or an element with `overflow: scroll` styling).
* Adds two props to `ADataTable`, which would enable infinite scrolling capabilities:
  * `onScrollToEnd: ({ inView, entry } : { inView: boolean, entry: IntersectionObserverEntry }) => void` - called when the user reaches the end of a data table's length
  * `maxHeight: string` - sets the maximum height of the data table to enable an overflow scroll

**What is the current behavior?** 

The `ADataTable` does not have functionality to support infinite scrolling.

**What is the new behavior (if this is a feature change)?**

Leverages the [Intersection Observer API]() in order to determine when the last row in the table is being viewed, which enables an external consumer of Atomic React to run a function that would (hypothetically) call the next batch of data table.

**Does this PR introduce a breaking change?** 

No

**Other Information**

* Associated PR for [eslint-config-atomic-react](https://github.com/cisco-sbg-ui/eslint-config-atomic-react)
* Please see [accessibility notes](https://atomic-react-19s81j0dz-securex.vercel.app/components/data-table#accessibility-notes) for infinite scrolling inside of a data table
* Also adds a new hook for using an [intersection observer](https://developer.mozilla.org/en-US/docs/Web/API/Intersection_Observer_API) in React
* Creates a way to enrich documentation examples without cluttering up the live code editor as seen on [Atomic React](https://atomic-react-19s81j0dz-securex.vercel.app/components/data-table#infinite-scrolling) component usage docs
* This work was initiated from [secure-client-forms](https://github.com/advthreat/secure-client-forms/issues/638)

### CODE REVIEW

File: framework/components/AInView/AInView.spec.js [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+    
+    it("should alert when an observed element toggles from being in and out of the view", () => {
+        cy.get(`${basicUsageSelector}`).get(`${basicUsageSelector} *[aria-live]`).contains("On Screen");
+        cy.get(`${basicUsageSelector} *[data-testid="scroll-container"]`).scrollTo("bottom");
+        cy.get(`${basicUsageSelector}`).get(`${basicUsageSelector} *[aria-live]`).contains("Off Screen");
+    });
+
+    it("should allow the toggler callback to run only once after first being in the view", () => {
+        cy.get(`${asyncUsageSelector}`).scrollIntoView();
+        cy.get(`${asyncUsageSelector} *[role="progressbar"]`).should("not.exist");
```
Comment (trent-boyd) [inline framework/components/AInView/AInView.spec.js:17]: More Cypress options:

```js
cy.get(`${asyncUsageSelector} *[role="progressbar"]`).as("progressBar").should("not.exist");
// other things here...
cy.get("@progressBar").should("be.visible");
```

File: framework/utils/hooks.js [hunk 2]
```diff
@@ -1,4 +1,4 @@
-import {useCallback, useEffect, useLayoutEffect, useRef, useState} from "react";
+import {useCallback, useMemo, useEffect, useLayoutEffect, useRef, useState} from "react";
```
Comment (trent-boyd) [inline framework/utils/hooks.js:1]: I got this far. Will review more later. :)

File: framework/components/ADataTable/ADataTable.js [hunk 3]
```diff
@@ -32,105 +60,128 @@ const ADataTable = forwardRef(
...
+                        onScrollToEnd(entry);
+                      }
+                    }}>
+                    {rowContent}
+                  </AInView>
+                );
+              })}
+            </tbody>
+          </ASimpleTable>
+        </ADataTableWrapper>
```
Comment (rwharrisjr) [inline framework/components/ADataTable/ADataTable.js:184]: Needs to have a default `return null` if the return conditions are falsy.

**Review by rwharrisjr — CHANGES_REQUESTED** [review]
Agreed with Trent's comments; lgtm otherwise.

**Review by rwharrisjr — APPROVED** [review]

**Pull request merged**